Feature/253 cancel unlocking

markuskreusch opened the pull request:
fixes #253

All unit tests are ok (did only a small test change for this) but I could not run the android tests. They fail with errors like `androidx.test.espresso.NoMatchingViewException: No views in hierarchy found matching: (with text: is "SDCARD" and has parent matching: with id is <2131362220> and is displayed on the screen to the user)` so I guess some setup is missing on my side.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/cryptomator/android?pullRequest=260) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/cryptomator/android?pullRequest=260) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/cryptomator/android?pullRequest=260) it.</sub>

SailReal reviewed (changes requested):
Hey @markuskreusch, great to hear from you and thanks for the PR :heart:!

Unfortunately, I noticed during testing that even though you followed the description in the issue 1:1, that if you unsubscribe from the `UnlockVaultUseCase`, the unlock continues to be performed in the background, which is to be expected but I didn't have in mind that this would lead to an partially unlocked vault. This can be seen on the one hand by the fact that the notification "Vaults unlocked: 1" appears and on the other hand, when the vault list is reloaded, for example by restarting the app, it shows that the vault is unlocked (but just partially).

I am currently thinking about the best way to solve this, but the unlock process should obviously be really cancelled...but thanks again for the PR :smiley: !!!

markuskreusch commented:
Mhm... didn't think about that but that is completely true.

The unlock background task per se is not cancelable or interruptible so even a new concept of a cancelable use case would not work. I suppose a way to transfer a cancellation status to the data layer and undo the unlock after it is done might work. Problem is that this will introduce a race condition so it is far from optimal and the only way to transfer a cancellation flag in a clean way would be another use case. Seems overly complicated.

Maybe the following could work:
Instead of unsubscribing the use case a a flag is added in the presenter, that tells if the unlock has been cancelled. If it is true the vault is immediately locked after it has been unlocked instead of navigating to the contents. I would have to check how this would affect the other presenters besides VaultListPresenter that use the unlock dialog though.

SailReal commented:
Yeah, I thought about an `CancellationException` but maybe the flag in the presenter is better and has less side effects. I did not think that this will be such an annoying construction but maybe we can still find a better solution :sweat_smile: ...

markuskreusch commented:
It should not be too much work so I will try to implement it this way and we can see if it is a suitable solution afterwards.

markuskreusch commented:
Seems to work well. Tested the process in the VaultListPresenter and the locking of the vault is working well. For a short while though the notification is displayed with one vault as unlocked. Not sure if this is acceptable.

Otherwise how exactly would your idea with a `CancellationException` work? Could not figure out a way that would effectively cancel the running use case but maybe I missed something.

SailReal commented:
Using the flag I could reproduce a few race conditions. One is easy to set up:
1. Unlock a (in my case local) vault.
2. Cancel the unlocking process. 
3. Loop this a few times. 
4. Unlock the vault. 

After the vault got unlocked, a few error-toasts are displayed
```java
org.cryptomator.debug E/ExceptionHandler: java.lang.IllegalStateException: Cryptor already registered for vault org.cryptomator.domain.Vault@1
        at org.cryptomator.data.cloud.crypto.CryptoCloudContentRepositoryFactory.registerCryptor(CryptoCloudContentRepositoryFactory.java:77)
        at org.cryptomator.data.cloud.crypto.CryptoCloudFactory.unlock(CryptoCloudFactory.java:86)
        at org.cryptomator.data.repository.CloudRepositoryImpl.unlock(CloudRepositoryImpl.java:102)
        at org.cryptomator.domain.usecases.vault.UnlockVault.execute(UnlockVault.java:26)
        at org.cryptomator.domain.usecases.vault.UnlockVaultUseCase$Launcher$2.call(UnlockVaultUseCase.java:101)
        at org.cryptomator.domain.usecases.vault.UnlockVaultUseCase$Launcher$2.call(UnlockVaultUseCase.java:95)
        at io.reactivex.internal.operators.flowable.FlowableFromCallable.subscribeActual(FlowableFromCallable.java:39)
        at io.reactivex.Flowable.subscribe(Flowable.java:14935)
        at io.reactivex.Flowable.subscribe(Flowable.java:14882)
        at io.reactivex.internal.operators.flowable.FlowableSubscribeOn$SubscribeOnSubscriber.run(FlowableSubscribeOn.java:82)
        at io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker$BooleanRunnable.run(ExecutorScheduler.java:288)
        at io.reactivex.internal.schedulers.ExecutorScheduler$ExecutorWorker.run(ExecutorScheduler.java:253)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:923)
    
    		ErrorCode: A5E6:86Q6
```

---

I thought we could use the [CancellationException](https://github.com/cryptomator/android/blob/develop/domain/src/main/java/org/cryptomator/domain/exception/CancellationException.java) and as in [UploadFiles#L37-L43](https://github.com/cryptomator/android/blob/develop/domain/src/main/java/org/cryptomator/domain/usecases/cloud/UploadFiles.java#L37-L43) a cancellation flag. This flag could be passed to `cloudRepository.unlock(...)` and changed from within the `UseCase` when `unlockVaultUseCase.cancel()` is called. So very similar to how we cancel the `UploadFilesUseCase` and the `CancelAwareStreams`.

Have it not yet 100% thought through whether this could actually be a solution but that is what I thought about with the `CancellationException`.

markuskreusch commented:
Mhm, we could solve the problem with the error toasts by unsubscribing the use case before restarting it. This could be considered kind of a bug because if a use case is invoked while it is running the original subscription is never unsubscribed. Remember that I saw that when I designed the use cases and never cared about it because usually we never invoked an already running use case again. Now it is a problem.

But maybe the cancellation is actually an option. I would design it the following way:

- The UnlockVault use case will get a cancelled flag, exactly like the UploadFiles use case
- It will pass this flag to CloudRepository.unlock which will pass it to CryptoCloudFactory.unlock
- The CryptoCloudFactory will check the cancellation flag before running line 86 and throw a CancellationException if required
- This way no cryptor will be registered if the use case has been cancelled and thus the vault remains locked
- It would be possible to add more checks of the flag and use some cancel aware stream to read the key file but from the users perspective this would not change anything besides saving the data volume for the keyfile download and some computation power for the kek derivation. I am not sure this would be worth the more complex code.

What do you think?

SailReal commented:
In my opinion, the variant with the real cancellations (a cancel flag, the `CancellationException` and without executing `CryptoCloudContentRepositoryFactory.registerCryptor(...)`) is the much cleaner option. Looks less complex, easier for the presenters to deal with the cancellations and less prone to race conditions etc.

> It would be possible to add more checks of the flag and use some cancel aware stream to read the key file but from the users perspective this would not change anything besides saving the data volume for the keyfile download and some computation power for the kek derivation. I am not sure this would be worth the more complex code.

In my opinion, the download of the `masterkey.cryptomator` file should not be interrupted, as it is only about 330 bytes in size and even with a poor mobile internet connection, a transfer of 330 bytes is almost impossible to interrupt. Interrupting the kek derivation is definitely much more attractive, but the cryptolib would have to be adapted and also be able to handle such a flag...I would say that not creating a cryptor should be enough to start with.

markuskreusch commented:
Ok, reverted the changes I did before and implemented it using a cancellation flag in the UnlockVaultUseCase. Works well IMHO and I could not reproduce race conditions any more except one small thing, which should not really be an issue:

If the user clicks on the cancel button just right before the unlock completes the EnterPasswordDialog will be closed by the press on cancel but a short time later the unlock completes and the app navigates to the vault list. It is only a small delay (I would guess 200ms or such, it required me a lot of tries to get that behaviour) so a user will probably not even notice that and just consider the click on cancel too late. In the end we can't do anything about it I suppose.

SailReal reviewed (changes requested):
Looks very clear now and I haven't found any hiccups in my few tests yet. Will test it tomorrow more intensively but it looks very good, thank you :clap: .

However, in any case, the changes regarding the DoLicenseCheck needs to be reverted :wink: 

> ...so a user will probably not even notice that and just consider the click on cancel too late. In the end we can't do anything about it I suppose.

Yes, I also think that fits. I'll have another look tomorrow, but as you say it can't be prevented completely anyway.

SailReal commented:
Even after more intensive testing, I was unable to detect any other abnormalities and as already mentioned, the code looks fine.
As soon as the desired changes have been made, this PR will be merged. Thank you for this contribution :pray:!

markuskreusch commented:
Ok, reverted the changes to DoLicenseCheck (and force pushed it to not keep the error in the history ;) ).
